File: leetcode/April-2020-challenge/04_Move_Zeroes.java
class Solution {
    public void moveZeroes(int[] nums) {
        int n = nums.length;
                     for(int i=0;i<n;i++){
                     for(int j=i+1;j<n;j++){
                            if(nums[i]==0){
                                int temp = nums[i];
                                nums[i]=nums[j];
                                nums[j]=temp;
                            }     
                       }
                   }    
        for(int j=0;j<n;j++){
           System.out.print(nums[j]);
        }
    }
}

// OR
class Solution {
    public void moveZeroes(int[] nums) {
        if(nums == null || nums.length ==0){
            return;
        }
        int length =0;
        for(int i = 0;i<nums.length;i++){
            if(nums[i] != 0){
                int temp = nums[i];
                nums[i] = nums[length];
                nums[length] = temp;
                length++;
            }

        }
    }
}

//OR


class Solution {
    public void moveZeroes(int[] nums) {
        if(nums==null || nums.length==0){
            return;
        }
        int left=0, i=0;
        int n=nums.length;
        while(i<nums.length){
            if(nums[i]!=0){
                swap(nums,left,i);
                left++;
            }
            i++;
        }
    }
    private void swap(int[] nums, int i, int j){
        int tmp=nums[i];
        nums[i]=nums[j];
        nums[j]=tmp;
    }
}

File: leetcode/April-2020-challenge/04_Move_Zeroes.cpp
/*
Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.

Example:

Input: [0,1,0,3,12]
Output: [1,3,12,0,0]
Note:

You must do this in-place without making a copy of the array.
Minimize the total number of operations.*/
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int n = nums.size();
        int nxt = 0;
        for(int x : nums) {
            if(x != 0) {
                nums[nxt] = x;
                nxt++;
            }
        }
        for(int i = nxt; i < n; i++) {
            nums[i] = 0;
        }   
        //return nums;
    }
};
© 2020 GitHub, In
